PR: https://github.com/odoo/odoo/pull/147668

From: b02faa688234f8e7ccaac211678d3a23e5f3764c
From: sofiagvaladze
Date: 2023-12-28 12:32:11

Structural Changes: 2
Total Changes: 33

[FIX] hr_holidays: fix activity creation for second approval

Prior, if the time off type needed second approval, there was no activity generated for the responsible.

This commit fixes that. Moreover we also organize the code by taking
references out of the loop.

task - 3584190

closes odoo/odoo#147668

X-original-commit: 5faa73427f6923f4d4d46d4b833c99405da9dfa2
Signed-off-by: Bertrand Dossogne (bedo) <bedo@odoo.com>
Signed-off-by: Sofie Gvaladze (sgv) <sgv@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/hr_leave.py
+++ b/addons/hr_holidays/models/hr_leave.py
@@ -1500,21 +1500,32 @@ Attempting to double-book your time off won't magically make your vacation 2x be
         return responsible
 
     def activity_update(self):
-        to_clean, to_do = self.env['hr.leave'], self.env['hr.leave']
+        to_clean, to_do, to_do_confirm_activity = self.env['hr.leave'], self.env['hr.leave'], self.env['hr.leave']
         activity_vals = []
+        today = fields.Date.today()
+        model_id = self.env.ref('hr_holidays.model_hr_leave').id
+        confirm_activity = self.env.ref('hr_holidays.mail_act_leave_approval')
+        approval_activity = self.env.ref('hr_holidays.mail_act_leave_second_approval')
         for holiday in self:
-            note = _(
-                'New %(leave_type)s Request created by %(user)s',
-                leave_type=holiday.holiday_status_id.name,
-                user=holiday.create_uid.name,
-            )
             if holiday.state == 'draft':
                 to_clean |= holiday
-            elif holiday.state == 'confirm':
+            elif holiday.state in ['confirm', 'validate1']:
                 if holiday.holiday_status_id.leave_validation_type != 'no_validation':
+                    if holiday.state == 'confirm':
+                        activity_type = confirm_activity
+                        note = _(
+                            'New %(leave_type)s Request created by %(user)s',
+                            leave_type=holiday.holiday_status_id.name,
+                            user=holiday.create_uid.name,
+                        )
+                    else:
+                        activity_type = approval_activity
+                        note = _(
+                            'Second approval request for %(leave_type)s',
+                            leave_type=holiday.holiday_status_id.name,
+                        )
+                        to_do_confirm_activity |= holiday
                     user_ids = holiday.sudo()._get_responsible_for_approval().ids or self.env.user.ids
-                    today = fields.Date.today()
-                    activity_type = self.env.ref('hr_holidays.mail_act_leave_approval')
                     for user_id in user_ids:
                         date_deadline = (
                             (holiday.date_from -
@@ -1529,7 +1540,7 @@ Attempting to double-book your time off won't magically make your vacation 2x be
                             'note': note,
                             'user_id': user_id,
                             'res_id': holiday.id,
-                            'res_model_id': self.env.ref('hr_holidays.model_hr_leave').id,
+                            'res_model_id': model_id,
                         })
             elif holiday.state == 'validate':
                 to_do |= holiday
@@ -1537,6 +1548,8 @@ Attempting to double-book your time off won't magically make your vacation 2x be
                 to_clean |= holiday
         if to_clean:
             to_clean.activity_unlink(['hr_holidays.mail_act_leave_approval', 'hr_holidays.mail_act_leave_second_approval'])
+        if to_do_confirm_activity:
+            to_do_confirm_activity.activity_feedback(['hr_holidays.mail_act_leave_approval'])
         if to_do:
             to_do.activity_feedback(['hr_holidays.mail_act_leave_approval', 'hr_holidays.mail_act_leave_second_approval'])
         self.env['mail.activity'].create(activity_vals)
